### ISSUE TITLE
feat: improve image cache logic.

### ISSUE THREAD

**Issue by justadogistaken** [alibaba/sealer#909]
## Issue Description
kubeadm join --cri-socket /var/run/cri-resmgr/cri-resmgr.sock 
![截屏2021-12-01 下午7 41 09](https://user-images.githubusercontent.com/38838049/144228544-dd854943-dcdd-428b-8d09-53109c7790c2.png)

Sealer hacked docker for supporting image cache and request endpoint redirecting.
I wonder if the scenes like "docker pull xxx" are rare. Basically the image pulling would triggered by kubelet in sealer.
If the above assumption is for sure. I think we can proxy image pulling request like cri-resource-manager(https://github.com/intel/cri-resource-manager), which changes kubelet cri-socket to their own process, and make it easier for intercepting image pulling requests. And it doesn't have to change any logic of docker.

**Comment by fanux** [comment]
I think integrate cri-resource-manager is an awesome idea. v0.7.0 version will try to use it.

**Comment by yj-cloud** [comment]
> I think integrate cri-resource-manager is an awesome idea. v0.7.0 version will try to use it.

All of images in the cluster are pulled from cached registry, the cached registry will take up more and more storage, it's a risk for master0 which the the cached registry is running. Pulling sealer cached images from the cached registry, rather than all of images in cluster, maybe it's better. Please think of it when integrating the cri-resource-manager.

**Comment by justadogistaken** [comment]
Find there is a flag "image-service-endpoint" for kubelet. Kubelet could seperate "image", "runtime" service. If we provide a grpc service to proxy "image" service only. The risk can be minimized.

**Comment by justadogistaken** [comment]
I write a simple grpc proxy server. And I did a test for change image name while pulling image.
1. pod yaml for image "nginx:1.14.2"
```
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - name: nginx
    image: nginx:1.14.2
    ports:
    - containerPort: 80
```

2. intercept pullImage request, and change the image name to `daocloud.io/library/nginx:1.9.1` in proxy server
3. kubectl create -f test.yaml
4. the pod started, and image is saved as `daocloud.io/library/nginx:1.9.1`
<img width="1140" alt="截屏2022-01-12 下午3 54 07" src="https://user-images.githubusercontent.com/38838049/149087050-d6398216-6c67-42ab-beab-ba253c7c03b4.png">

**Comment by fanux** [comment]
@justadogistaken cool! PR the code~

**Comment by justadogistaken** [comment]
I‘ve submitted code here: https://github.com/osemp/sealer-cri-shim.
Just search "TODO", that is where we need to be aware.

**Comment by justadogistaken** [comment]
> I‘ve submitted code here: https://github.com/osemp/sealer-cri-shim. Just search "TODO", that is where we need to be aware of.

change “image-service-endpoint” of kubelet to make it work.

**Comment by shy-Xu** [comment]
> image

What kind of container runtime did you use? docker or containerd?

**Comment by shy-Xu** [comment]
> I write a simple grpc proxy server. And I did a test for changing image name while pulling image.
> 
> 1. pod yaml for image "nginx:1.14.2"
> 
> ```
> apiVersion: v1
> kind: Pod
> metadata:
>   name: nginx
> spec:
>   containers:
>   - name: nginx
>     image: nginx:1.14.2
>     ports:
>     - containerPort: 80
> ```
> 
> 2. intercept pullImage request, and change the image name to `daocloud.io/library/nginx:1.9.1` in proxy server
> 3. kubectl create -f test.yaml
> 4. the pod started, and image is saved as `daocloud.io/library/nginx:1.9.1`
> 
> <img alt="截屏2022-01-12 下午3 54 07" width="1140" src="https://user-images.githubusercontent.com/38838049/149087050-d6398216-6c67-42ab-beab-ba253c7c03b4.png">

What kind of container runtime did you use? docker or containerd?

**Comment by justadogistaken** [comment]
> > I write a simple grpc proxy server. And I did a test for changing image name while pulling image.
> > 
> > 1. pod yaml for image "nginx:1.14.2"
> > 
> > ```
> > apiVersion: v1
> > kind: Pod
> > metadata:
> >   name: nginx
> > spec:
> >   containers:
> >   - name: nginx
> >     image: nginx:1.14.2
> >     ports:
> >     - containerPort: 80
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 2. intercept pullImage request, and change the image name to `daocloud.io/library/nginx:1.9.1` in proxy server
> > 3. kubectl create -f test.yaml
> > 4. the pod started, and image is saved as `daocloud.io/library/nginx:1.9.1`
> > 
> > <img alt="截屏2022-01-12 下午3 54 07" width="1140" src="https://user-images.githubusercontent.com/38838049/149087050-d6398216-6c67-42ab-beab-ba253c7c03b4.png">
> 
> What kind of container runtime did you use? docker or containerd?

I used docker, but I think it's irrelevant with the concrete cri. Just change the “image-service-endpoint” of kubelet to our cri-shim, and make the kubelet fetch image through our cri-shim. And add some logics in our cri-shim meanwhile, then let cri-shim invoke the running cri socket eventually.

**Comment by shy-Xu** [comment]
Right, I asked this because I find containerd is used as container runtime at [cri-resource-manager](https://intel.github.io/cri-resource-manager/stable/docs/quick-start.html#pre-requisites).  I tried the [quick start](https://intel.github.io/cri-resource-manager/stable/docs/quick-start.html#) with docker but failed. So I am eager to know how you achieved it, and I am thinking of how to integrate it to sealer.

**Comment by justadogistaken** [comment]
> Right, I asked this because I find containerd is used as container runtime at [cri-resource-manager](https://intel.github.io/cri-resource-manager/stable/docs/quick-start.html#pre-requisites). I tried the [quick start](https://intel.github.io/cri-resource-manager/stable/docs/quick-start.html#) with docker but failed. So I am eager to know how you achieved it, and I am thinking of how to integrate it to sealer.

Do use "cri-resource-manager" directly, it has more abilities than what we need, and it may has some problems, we just need to build a gRPC bridge between cri and kubelet. If you want to use docker, try to shim for `unix:///var/run/dockershim.sock`(is the name correct? not the dockerd.sock anyway). Note: not to use cri-resource-manager directly, it has many features that we have not tested.

**Comment by shy-Xu** [comment]
Use CloudImage `kubernetes-crishim:v1.19.8` to experience this feature.  `kubernetes-crishim:v1.19.8` changes images without specified registry to sea.hub:5000/<image-path>:<image-tag>, images with specified registry will not be processed

**Comment by cuisongliu** [comment]
<img width="1121" alt="image" src="https://user-images.githubusercontent.com/4150905/152724886-00bbf0b4-ce96-454f-a67f-aa599aef62ef.png">
domain=="" to replace private ????

**Comment by shy-Xu** [comment]
> <img alt="image" width="1121" src="https://user-images.githubusercontent.com/4150905/152724886-00bbf0b4-ce96-454f-a67f-aa599aef62ef.png">
> 
> domain=="" to replace private ????

To replace DockerHub images

**Comment by cuisongliu** [comment]
> > <img alt="image" width="1121" src="https://user-images.githubusercontent.com/4150905/152724886-00bbf0b4-ce96-454f-a67f-aa599aef62ef.png">
> > domain=="" to replace private ????
> 
> To replace DockerHub images

domain!="" replace it. this logic is has some propleme
